docs/L1-L2 Communication/messaging-mechanism.md

```diff
@@ -42,3 +53,14 @@ The above flow is illustrated in the following diagram:
 Imagine a scenario where a user transfers an asset from L1 to L2. The flow starts with the user sending the asset to a StarkNet bridge and the corresponding L1→L2 message generation. Now, imagine that the L2 message consumption doesn’t function (this might happen due to a bug in the dApps’s Cairo contract). This could result in the user losing custody over their asset forever.
 
 To mitigate this risk, we allow the contract that initiated the L1→L2 message to cancel it – after declaring the intent and waiting a suitable amount of time. The user starts by calling [`startL1ToL2MessageCancellation`](https://github.com/starkware-libs/cairo-lang/blob/4e233516f52477ad158bc81a86ec2760471c1b65/src/starkware/starknet/eth/StarknetMessaging.sol#L134) with the relevant message parameters in the StarkNet Core Contract. Then, after a five days delay, the user can finalize the cancellation by calling [`cancelL1ToL2Message`](https://github.com/starkware-libs/cairo-lang/blob/4e233516f52477ad158bc81a86ec2760471c1b65/src/starkware/starknet/eth/StarknetMessaging.sol#L147). The reason for the delay is to protect the sequencer from a DOS attack in the form of repeatedly sending and canceling a message – before it is included in L1, rendering the L2 block which contains the activation of the corresponding L1 handler invalid. Note that this flow should only be used in edge cases such as bugs on the Layer 2 contract preventing message consumption.
+
+## L1 → L2 Message Fees
+
+An L1 → L2 message induces a transaction on L2 which is not associated with an account (unlike regular transactions). This calls for a different mechanism for paying the transaction's fee, for otherwise the sequencer has no incentive of including l1 handler calls inside a block.
+
+To avoid having to interact with both L1, L2 when sending a message, L1 → L2 messages are paypable on L1, by sending ETH with the call to the payable function `sendMessageToL2` on the StarkNet Core contract. This payment will be taken as a fee for handling the message, and will be charged by the sequencer in full upon updating the L1 state with the consumption of this message.
```

Comment:
```suggestion
To avoid having to interact with both L1 and L2 when sending a message, L1 → L2 messages are payable on L1, by sending ETH with the call to the payable function `sendMessageToL2` on the StarkNet Core contract. The sequencer takes this fee in exchange for handling the message. The sequencer charges the fee in full upon updating the L1 state with the consumption of this message.
```